File: RSA.cpp
#include <iostream>
#include <cmath>

using namespace std;


int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}


long long modExpo(long long base, long long exp, long long mod) {
    long long result = 1;
    while (exp > 0) {
        if (exp % 2 == 1) {
            result = (result * base) % mod;
        }
        base = (base * base) % mod;
        exp /= 2;
    }
    return result;
}


int modInverse(int e, int phi) {
    int t1 = 0, t2 = 1, r1 = phi, r2 = e;
    while (r2 != 0) {
        int quotient = r1 / r2;
        int temp = t1;
        t1 = t2;
        t2 = temp - quotient * t2;

        temp = r1;
        r1 = r2;
        r2 = temp - quotient * r2;
    }

    if (t1 < 0) {
        t1 += phi;
    }
    return t1;
}

int main() {
    
    int p = 61; 
    int q = 53; 

    
    int n = p * q;

    
    int phi = (p - 1) * (q - 1);

    
    int e = 17; 
    if (gcd(e, phi) != 1) {
        cout << "e and phi are not coprime. Choose a different value for e." << endl;
        return -1;
    }

    
    int d = modInverse(e, phi);

    cout << "Public Key (e, n): (" << e << ", " << n << ")" << endl;
    cout << "Private Key (d, n): (" << d << ", " << n << ")" << endl;

    
    int message;
    cout << "Enter the message to encrypt (as a number): ";
    cin >> message;

    
    long long encryptedMessage = modExpo(message, e, n);
    cout << "Encrypted Message: " << encryptedMessage << endl;

    
    long long decryptedMessage = modExpo(encryptedMessage, d, n);
    cout << "Decrypted Message: " << decryptedMessage << endl;

    return 0;
}

File: java_codes/RSA.java
import java.math.BigInteger;

public class RSA {

    // Function to calculate gcd
    public static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    // Function to compute modular inverse using Extended Euclidean Algorithm
    public static int modInverse(int a, int m) {
        int m0 = m;
        int y = 0, x = 1;

        if (m == 1) return 0;

        while (a > 1) {
            int q = a / m;
            int t = m;

            m = a % m;
            a = t;
            t = y;

            y = x - q * y;
            x = t;
        }

        if (x < 0) x += m0;

        return x;
    }

    // Function for modular exponentiation
    public static int modPow(int base, int exp, int mod) {
        int result = 1;
        base = base % mod;
        
        while (exp > 0) {
            if ((exp % 2) == 1) {
                result = (result * base) % mod;
            }
            exp = exp >> 1; // exp /= 2
            base = (base * base) % mod;
        }
        return result;
    }

    public static void main(String[] args) {
        int p = 7;
        int q = 11;
        int n = p * q; // Modulus
        int phi = (p - 1) * (q - 1); // Euler's totient function
        int e = 13; // Public exponent

        // Check if e is coprime with phi
        while (e < phi) {
            if (gcd(e, phi) == 1) break;
            e++;
        }

        int d = modInverse(e, phi); // Private exponent

        System.out.println("Public key (n, e): (" + n + ", " + e + ")");
        System.out.println("Private key (d): " + d);

        int msg = 5; 
        System.out.println("Original Message: " + msg);

        // Encryption
        int c = modPow(msg, e, n);
        System.out.println("Encrypted data: " + c);

    
        int pt = modPow(c, d, n);
        System.out.println("Decrypted (Original Message): " + pt);
    }
}
